### ISSUE TITLE
UX of tripped `assert`s in `const` invocations could be a lot better

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/sdk#55485]
Consider this example:

```dart
import 'package:meta/meta.dart';

void main() {
  const red = Color(0xFFFF0000);
  const tooSmall = Color(-1);
  const tooBig = Color(0xFFFFFFFF + 1);
}

extension type const Color._(int _argb) {
  const Color(@mustBeConst this._argb)
      : assert(_argb >= 0 && _argb <= 0xFFFFFFFF, 'Out of range: $_argb');
}
```

Both `tooSmall` and `tooBig` are correctly flagged by the analyzer (yay!), but the errors look like this:

```txt
Evaluation of this constant expression throws an exception.dart(const_eval_throws_exception)
foo.dart(11, 9): The exception is 'The assertion in this constant expression failed.' and occurs here.
```

Unless it's somehow impossible, it would be much nicer to put the actual assert message into the error:

```txt
foo.dart(11, 9): Out of range: -1
```

**Comment by bwilkerson** [comment]
I agree, that would be better. It won't be possible in some kinds of invalid code, but if the code is valid then we ought to be able to do that.

@kallentu

**Comment by matanlurey** [comment]
I assume you're asking Brian but I think that would be great!

**Comment by bwilkerson** [comment]
I don't think that message is too verbose.

I might tweak the wording a bit, making it more general and slightly shorter. Maybe something like "An assertion failed with the message '{0}'." It would also be good if there were a context message pointing to the location of the assertion, given that it's likely to be distant from the expression where the diagnostic is being reported.

**Comment by matanlurey** [comment]
Really appreciate you both improving this!

**Comment by kallentu** [comment]
> Really appreciate you both improving this!

Thanks for filing the bug! I wouldn't have thought about it otherwise.

But should be improved now with https://github.com/dart-lang/sdk/commit/26652c6d31f304b235e593f98d1c7431e6e9207c, so closing.
